### PR TITLE
Increase functionality example app

### PR DESCRIPTION
**Pull request by remonh87**
Added logs and discovered services to example app

https://user-images.githubusercontent.com/20230599/109391653-6ecc7080-7918-11eb-9a8a-3d51f9d267f6.mov

### CODE REVIEW

**Comment by remonh87** [comment]
Next steps is adding read write and subscribe

File: example/lib/main.dart [hunk 1]
```diff
@@ -1,27 +1,35 @@
...
+  final _serviceDiscoverer =
+      BleServiceDiscoverer(_ble.discoverServices, _bleLogger.addToLog);
   runApp(
     MultiProvider(
       providers: [
         Provider.value(value: _scanner),
         Provider.value(value: _monitor),
         Provider.value(value: _connector),
+        Provider.value(value: _serviceDiscoverer),
+        Provider.value(value: _bleLogger),
```
Comment (werediver) [inline example/lib/main.dart:32]: If you are deep in refactoring this, you may want to look into the successor, of sorts, to the "provider" package: [rrousselGit/river_pod](https://github.com/rrousselGit/river_pod). The author of "provider" attempts to overcome some of the flaws of its previous creation. I haven't tried it [yet], but am curious.

File: example/lib/src/ble/ble_logger.dart [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+class BleLogger {
+  final List<String> _logMessages = [];
+  final DateFormat formatter = DateFormat('hh:mm:ss');
+
+  List<String> get messages => _logMessages;
+
+  void addToLog(String message) {
+    final now = DateTime.now();
+    _logMessages.add('${formatter.format(now)}.${now.millisecond} - $message');
```
Comment (werediver) [inline example/lib/src/ble/ble_logger.dart:11]: Couldn't you use the date formatter to render fractions of seconds. I believe, it must be something like `hh:mm:ss.SSS`. Try it of check the spec.

File: example/lib/src/ble/ble_scanner.dart [hunk 3]
```diff
@@ -5,9 +5,10 @@ import 'package:flutter_reactive_ble_example/src/ble/reactive_state.dart';
 import 'package:meta/meta.dart';
 
 class BleScanner implements ReactiveState<BleScannerState?> {
-  BleScanner(this._ble);
+  BleScanner(this._ble, this._logMessage);
```
Comment (werediver) [inline example/lib/src/ble/ble_scanner.dart:8]: Perhaps, names parameters would be better here.

File: example/lib/src/ui/device_detail/device_detail_screen.dart [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import 'package:flutter_reactive_ble_example/src/ble/ble_device_connector.dart';
+import 'package:flutter_reactive_ble_example/src/ui/device_detail/device_log_tab.dart';
+import 'package:provider/provider.dart';
+
+import 'device_interaction_tab.dart';
+
+class DeviceDetailScreen extends StatelessWidget {
+  final DiscoveredDevice device;
+
+  const DeviceDetailScreen({required this.device, Key? key}) : super(key: key);
```
Comment (werediver) [inline example/lib/src/ui/device_detail/device_detail_screen.dart:12]: Wow, no @ 😎 Wow², nullability 😍

File: lib/src/discovered_devices_registry.dart [hunk 5]
```diff
@@ -40,7 +40,8 @@ class DiscoveredDevicesRegistryImpl implements DiscoveredDevicesRegistry {
   bool isEmpty() => discoveredDevices.isEmpty;
 
   @override
-  bool deviceIsDiscoveredRecently({String? deviceId, Duration? cacheValidity}) =>
+  bool deviceIsDiscoveredRecently(
+          {String? deviceId, Duration? cacheValidity}) =>
```
Comment (werediver) [inline lib/src/discovered_devices_registry.dart:44]: Would look better with a trailing comma.

File: lib/src/generated/bledata.pbjson.dart [hunk 6]
```diff
@@ -195,16 +368,44 @@ const DiscoverServicesInfo$json = const {
...
+      '10': 'characteristicUuids'
+    },
+    const {
+      '1': 'includedServices',
+      '3': 3,
+      '4': 3,
+      '5': 11,
+      '6': '.DiscoveredService',
+      '10': 'includedServices'
+    },
```
Comment (werediver) [inline lib/src/generated/bledata.pbjson.dart:408]: I kind of hate Protobuf. But I love that it synchronizes multiple parties (languages) together, so it's a love-hate relationship 🤔

**Review by werediver — APPROVED** [review]
Check the comments (most of them are really minor) and :shipit:

File: example/lib/main.dart (hunk 1 above)
Comment (remonh87) [reply example/lib/main.dart:32]: yess I really like rived pod but decided to not go for refactoring at the moment but maybe in a next pr :)

File: lib/src/generated/bledata.pbjson.dart (hunk 6 above)
Comment (remonh87) [reply lib/src/generated/bledata.pbjson.dart:408]: same here :-)

**Pull request merged**